training/tf/convert_to_v3.py

```diff
@@ -0,0 +1,213 @@
+        self.flat_planes = []
+        for i in range(256):
+            self.flat_planes.append(bytes([i]*64))
+        self.init_structs()
+        self.history = []
+        for history in range(NUM_HIST):
+            self.history.append(Board())
+        self.us_ooo = 0
+        self.us_oo  = 0
+        self.them_ooo = 0
+        self.them_oo  = 0
+        self.us_black = 0
+        self.rule50_count = 0
+
+    def init_structs(self):
+        # struct.Struct doesn't pickle, so it needs to be separately
+        # constructed in workers.
+
+        # V2 Format (8604 bytes total)
+        # int32 version (4 bytes)
+        # 1924 float32 probabilities (7696 bytes)
+        # 112 packed bit planes (896 bytes)
+        # uint8 castling us_ooo (1 byte)
+        # uint8 castling us_oo (1 byte)
+        # uint8 castling them_ooo (1 byte)
+        # uint8 castling them_oo (1 byte)
+        # uint8 side_to_move (1 byte)
+        # uint8 rule50_count (1 byte)
+        # uint8 move_count (1 byte)
+        # int8 result (1 byte)
+        self.v2_struct = struct.Struct('4s7696s896sBBBBBBBb')
+
+    def clear_hist(self):
+        for hist in range(NUM_HIST):
+            self.history.clear_board()
+
+    def update_board(self, hist, piece, hex_string):
+        """
+            Update the ASCII board representation
+        """
+        bit_board = int(hex_string, 16)
+        for r in range(8):
+            for f in range(8):
+                if bit_board & (1<<(r*8+f)):
+                    assert(self.history[hist].board[r][f] == ".")
+                    self.history[hist].board[r][f] = piece
+                    #self.history[hist].board[r] = \
+                    #    self.history[hist].board[r][:f] + \
+                    #    piece + \
+                    #    self.history[hist].board[r][8-f:]
+                    
+    def describe(self):
+        s = ""
+        if self.us_black:
+            s += "us = Black\n"
+        else:
+            s += "us = White\n"
+        s += "rule50_count {} b_ooo b_oo, w_ooo, w_oo {} {} {} {}\n".format(
+            self.rule50_count, self.us_ooo, self.us_oo, self.them_ooo, self.them_oo)
+        rank_strings = []
+        for hist in range(NUM_HIST):
+            rank_strings.append(self.history[hist].describe())
+        for rank in range(8):
+            for hist in range(NUM_HIST):
+                s += rank_strings[hist][rank] + " "
+            s += "\n"
+        return s
+
+    def update_reals(self, text_item):
+        self.us_ooo = int(text_item[NUM_HIST*NUM_PLANES+0])
+        self.us_oo = int(text_item[NUM_HIST*NUM_PLANES+1])
+        self.them_ooo =int(text_item[NUM_HIST*NUM_PLANES+2])
+        self.them_oo = int(text_item[NUM_HIST*NUM_PLANES+3])
+        self.us_black =int(text_item[NUM_HIST*NUM_PLANES+4])
+        self.rule50_count = min(int(text_item[NUM_HIST*NUM_PLANES+5]), 255)  # TODO should be around 99-102ish
+
+    def convert_v1_to_v3(self, ply, text_item):
```

Comment:
I think it's better to use the chunked here and convert from binary format. chunkparser handles conversion from V1 for free. V1 -> V2 is well tested in chunkparser.